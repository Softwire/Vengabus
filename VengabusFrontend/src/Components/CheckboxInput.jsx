import React, { Component } from 'react';
import { css } from 'emotion';
import { Checkbox, Glyphicon, OverlayTrigger } from 'react-bootstrap';

/**
 * @prop {bool} data The data that is set by the checkbox.
 * @prop {node} tooltip The tooltip element that is displayed when hovering over the info glyphicon. If undefined then no info tooltip.
 * @prop {function} onChange The function that is called when the value of the checkbox is changed.
 * @prop {string} text The name of the property that is being set by the checkbox.
 */
export class CheckboxInput extends Component {

    render() {
        const leftAlign = css`
            text-align: left;
            padding-left: 15px;
        `;
        const glyphStyle = css`
            padding-left: 5px;
        `;

        return (
            <div>
                {
                    this.props.tooltip ? (
                        <p className={leftAlign}>
                            {this.props.text}
                            <OverlayTrigger placement="right" overlay={this.props.tooltip}>
                                <Glyphicon glyph="info-sign" className={glyphStyle} />
                            </OverlayTrigger>
                        </p>
                    ) : (
                            <p className={leftAlign}>{this.props.text}</p>
                        )
                }
                <Checkbox
                    className={leftAlign}
                    checked={this.props.data}
                    onChange={(event) => {
                        this.props.onChange(event.target.checked);
                    }}
                />
            </div>
        );
    }

}


