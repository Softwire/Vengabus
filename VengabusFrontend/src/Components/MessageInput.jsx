import React, { Component } from 'react';
import { css } from 'react-emotion';
import { MessagePropertyInput } from './MessagePropertyInput';
import { serviceBusConnection } from '../AzureWrappers/ServiceBusConnection';
import {
    FormGroup,
    FormControl,
    ControlLabel,
    Button
} from "react-bootstrap";
import Select from 'react-select';
import _ from 'lodash';
const classNames = require('classnames');

/** 
 * @prop {Object} data All of the data that is passed down into this component.
 * @property {Object} data.message Can take a message as a prop to replay message.
 * @property {Object} data.message.customProperties User defined properties (key-string pairs).
 * @property {string} data.message.MessageBody The text of the message.
 * @property {string} data.message.MessageId The ID of the message.
 * @property {string} data.message.predefinedProperties Any other predefined properties used by Azure.
 * @property {bool} data.recipientIsQueue Is the recepient of the message a queue?
 * @property {string} data.selectedQueue Name of the recepient queue.
 * @property {string} data.selectedTopic Name of the recepient topic.
 */

export class MessageInput extends Component {
    constructor(props) {
        super(props);
        const data = this.props.data || {};
        const message = data.message;
        this.arePredefinedPropsLoaded = false;
        this.state = {
            permittedValues: [],
            availableTopics: [],
            availableQueues: [],
            recipientIsQueue: data.recipientIsQueue ? data.recipientIsQueue : true,
            messageBody: message ? message.MessageBody : '',
            userDefinedProperties: message ? this.getUserDefinedProperties(message) : [], //[{name: something, value: something}]
            preDefinedProperties: [], //Not set here because permitted values must be fetched first
            reservedPropertyNames: [], //a list of name of possible readable properties of a message
            selectedQueue: data.selectedQueue ? data.selectedQueue : undefined,
            selectedTopic: data.selectedTopic ? data.selectedTopic : undefined
        };
    }

    componentDidMount() {
        this.serviceBusService = serviceBusConnection.getServiceBusService();
        this.serviceBusService.getWriteableMessageProperties().then((result) => {
            this.arePredefinedPropsLoaded = true;
            this.setState({
                permittedValues: result,
                preDefinedProperties: this.props.message ? this.getPreDefinedProperties(this.props.message) : [] //[{name: something, value: something}]
            });
        });
        this.serviceBusService.listQueues().then((result) => {
            this.setState({
                availableQueues: _(result.data).map((queue) => { return { value: queue.name, label: queue.name }; })
            });
        });
        this.serviceBusService.listTopics().then((result) => {
            this.setState({
                availableTopics: _(result.data).map((topic) => { return { value: topic.name, label: topic.name }; })
            });
        });

        this.serviceBusService.getReadableMessageProperties().then((result) => {
            this.setState({
                reservedPropertyNames: result
            });
        });
    }

    getUserDefinedProperties = (message) => {
        return this.getTargetProperties(message, "customProperties");
    }

    getPreDefinedProperties = (message) => {
        return this.getTargetProperties(message, "predefinedProperties");
    }

    getTargetProperties = (message, propertyClass) => {
        const properties = [];
        const keys = Object.keys(message[propertyClass]);
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            properties.push({
                name: key,
                value: message[propertyClass][key]
            });
        }
        return properties;
    }

    /**
     * Updates a user-defined property name or value in the state.
     * @param {integer} position The position of the property in the list.
     * @param {string} attribute The attribute of the property that has changed, 'name' or 'value' 
     * @param {string} newValue The new value of that attribute of the property.
     */
    handleUserDefinedPropertyChange = (position, attribute, newValue) => {
        const newUserDefinedProperties = [...this.state.userDefinedProperties];
        newUserDefinedProperties[position][attribute] = newValue;
        this.setState({ userDefinedProperties: newUserDefinedProperties });
    };

    /**
    * Updates a pre-defined property name or value in the state.
    * @param {integer} position The position of the property in the list.
    * @param {string} attribute The attribute of the property that has changed, 'name' or 'value' 
    * @param {string} newValue The new value of that attribute of the property.
    */
    handlePreDefinedPropertyChange = (position, attribute, newValue) => {
        let newPreDefinedProperties = [...this.state.preDefinedProperties];
        newPreDefinedProperties[position][attribute] = newValue;
        this.setState({ preDefinedProperties: newPreDefinedProperties });
    };

    /**
     * Updates a collection of properties by applying the given updateOperation to it.
     * @param {boolean} isUserDefined Should be true if the property is user-defined, false if it is a pre-defined property.
     * @param {funciton} updateOperation The operation to be applied to the properties collection.
     */
    updatePropertiesCollection = (isUserDefined, updateOperation) => {
        const propertyType = isUserDefined ? "userDefinedProperties" : "preDefinedProperties";
        const newProperties = [...this.state[propertyType]];
        updateOperation(newProperties);
        this.setState({
            //https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names
            [propertyType]: newProperties
        });
    };

    /**
     * Adds a new property to the list of user-defined properties.
     * @param {boolean} isUserDefined Should be true if the property is user-defined, false if it is a pre-defined property.
     */
    addNewProperty = (isUserDefined) => {
        this.updatePropertiesCollection(isUserDefined, (propertyCollectionToMutate) => {
            propertyCollectionToMutate.push({ name: "", value: "" });
        });
    }

    /**
     * Deletes a row from the list of user defined properties.
     * @param {integer} index The index of the row to delete.
     * @param {boolean} isUserDefined Should be true if the property is user-defined, false if it is a pre-defined property.
     */
    deleteRow = (index, isUserDefined) => {
        this.updatePropertiesCollection(isUserDefined, (propertyCollectionToMutate) => {
            propertyCollectionToMutate.splice(index, 1);
        });
    };

    /**
     * Updates the message body in the state with a new value.
     * @param {string} newBody The new value of the body.
     */
    handleMessageBodyChange = newBody => {
        this.setState({ messageBody: newBody });
    };

    /**
     * Changes which queue or topic the message will be sent to.
     * @param {string} newName The name of the queue or topic.
     */
    handleQueueOrTopicChange = (newName) => {
        if (this.state.recipientIsQueue) {
            this.setState({
                selectedQueue: newName,
                selectedTopic: undefined
            });
        } else {
            this.setState({
                selectedQueue: undefined,
                selectedTopic: newName
            });
        }
    }

    /**
     * Changes whether the message will be sent to a queue or topic.
     * @param {boolean} isQueue True if the message should be send to a queue, false if it should be sent to a topic.
     */
    handleRecipientTypeChange = (isQueue) => {
        this.setState({
            recipientIsQueue: isQueue
        });
    }

    /**
     * Converts an array of properties to a properties object:
     * [{name: name1, value: value1}] -> {name1: value1}
     * @param {object[]} properties The array of properties to convert.
     * @returns {object} The created properties object.
     */
    createMessagePropertyDictionary = (properties) => {
        const ret = {};
        for (let i = 0; i < properties.length; i++) {
            const thisPropertyName = properties[i].name;
            const thisPropertyValue = properties[i].value;
            //Prevent the user from inputting invalid property names.
            //Cannot use isPropertyNameInvalid here because if there are two properties with the same name it will mark
            //both of them as invalid whereas we just want to remove one of them.
            if (thisPropertyName && thisPropertyValue && !ret.hasOwnProperty(thisPropertyName)) {
                if (thisPropertyName.length > 0) {
                    ret[thisPropertyName] = thisPropertyValue;
                }
            }
        }
        return ret;
    }

    /**
     * Converts the userDefinedProperties and preDefinedProperties arrays to a single message object
     * in the format that is accepted by the API.
     * @returns {object} The created message.
     */
    createMessageObject = () => {
        const message = {};
        message.customProperties = this.createMessagePropertyDictionary(this.state.userDefinedProperties);
        message.predefinedProperties = this.createMessagePropertyDictionary(this.state.preDefinedProperties);
        message.MessageBody = this.state.messageBody;
        return message;
    }

    /**
     * Clears the message body and removes all properties from the message.
     */
    discardMessage = () => {
        this.setState({
            messageBody: '',
            userDefinedProperties: [],
            preDefinedProperties: []
        });
    }

    /**
     * Sends the message to the selected queue/topic.
     */
    submit = () => {
        const message = this.createMessageObject();
        if (this.state.recipientIsQueue) {
            this.serviceBusService.sendMessageToQueue(this.state.selectedQueue, message);
        } else {
            this.serviceBusService.sendMessageToTopic(this.state.selectedTopic, message);
        }
    }

    render() {
        const formStyle = css`
            margin-left: 5px;
            margin-right: 5px;
            padding-top: 1%;
            width: calc(100% - 10px); /* 10px total margin */
            float: left;
        `;
        const queueOrTopicSelectionRadioStyle = css`
            float: left;
            width: 2%;
            position:relative;
            top:9px;
        `;
        const queueOrTopicSelectionStyle = css`
            float: left;
            width: 5%;
        `;
        const queueOrTopicSelectionDropdownStyle = css`
            float: left;
            width: 92%;
        `;
        const buttonStyle = css`
            width: 270px;
            margin-left: 5px;
        `;
        const headingStyle = css`
            font-weight: bold;
            margin-left: 5px;
        `;
        const leftAlignContainerStyle = css`
            text-align:left;
        `;
        const bodyStyle = css`
            min-height: 350px;
            padding-left: 5px;
        `;
        const verticalAlign = css`
            line-height: 38px;
        `;
        const fullWidth = css`
            float: left;
            width: 100%;
        `;
        const buttonLoading = css`
            opacity: 0.5;
            :hover {
                cursor: progress;
            }
        `;
        let preDefinedPropsButtonClassNames = classNames(buttonStyle, this.arePredefinedPropsLoaded || buttonLoading);
        const preDefinedPropertiesButtonText = this.arePredefinedPropsLoaded ? 'Add new Azure property' : 'Loading pre-defined properties...';
        return (
            <div className={formStyle} >
                <div className={leftAlignContainerStyle}>
                    <p className={headingStyle}>Destination</p>
                </div>
                <div className={fullWidth}>
                    <div
                        onClick={() => this.handleRecipientTypeChange(true)}
                    >
                        <input
                            className={queueOrTopicSelectionRadioStyle}
                            type="radio"
                            value="queue"
                            checked={this.state.recipientIsQueue}
                            onChange={() => this.handleRecipientTypeChange(true)}
                        />
                        <div className={leftAlignContainerStyle + " " + queueOrTopicSelectionStyle + " " + headingStyle + " " + verticalAlign}>
                            <p>Queue</p>
                        </div>
                    </div>
                    <Select
                        isDisabled={!this.state.recipientIsQueue}
                        className={queueOrTopicSelectionDropdownStyle}
                        title="Queue"
                        id={"queue-dropdown"}
                        options={this.state.availableQueues}
                        value={this.state.selectedQueue ? { value: this.state.selectedQueue, label: this.state.selectedQueue } : undefined}
                        onChange={(event) => this.handleQueueOrTopicChange(event.value)}
                    />
                </div>
                <div className={fullWidth}>
                    <div
                        onClick={() => this.handleRecipientTypeChange(false)}
                    >
                        <input
                            className={queueOrTopicSelectionRadioStyle}
                            type="radio"
                            value="topic"
                            checked={!this.state.recipientIsQueue}
                            onChange={() => this.handleRecipientTypeChange(false)}
                        />
                        <div className={leftAlignContainerStyle + " " + queueOrTopicSelectionStyle + " " + headingStyle + " " + verticalAlign}>
                            <p>Topic</p>
                        </div>
                    </div>
                    <Select
                        isDisabled={this.state.recipientIsQueue}
                        className={queueOrTopicSelectionDropdownStyle}
                        title="Topic"
                        id={"topic-dropdown"}
                        options={this.state.availableTopics}
                        value={this.state.selectedTopic ? { value: this.state.selectedTopic, label: this.state.selectedTopic } : undefined}
                        onChange={(event) => this.handleQueueOrTopicChange(event.value)}
                    />
                </div>
                <hr className={fullWidth} />
                <div className={leftAlignContainerStyle}>
                    <p className={headingStyle}>Pre-defined Properties</p>
                </div>
                <MessagePropertyInput
                    properties={this.state.preDefinedProperties}
                    handlePropertyNameChange={(newName, index) => this.handlePreDefinedPropertyChange(index, 'name', newName)}
                    handlePropertyValueChange={(newValue, index) => this.handlePreDefinedPropertyChange(index, 'value', newValue)}
                    deleteRow={(index) => this.deleteRow(index, false)}
                    permittedValues={this.state.permittedValues}
                />
                <form>
                    <div className={leftAlignContainerStyle}>
                        <Button
                            className={preDefinedPropsButtonClassNames}
                            onClick={() => this.addNewProperty(false)}
                        >
                            {preDefinedPropertiesButtonText}
                        </Button>
                    </div>
                </form>
                <hr />
                <div className={leftAlignContainerStyle}>
                    <p className={headingStyle}>User-defined Properties</p>
                </div>
                <MessagePropertyInput
                    properties={this.state.userDefinedProperties}
                    handlePropertyNameChange={(newName, index) => this.handleUserDefinedPropertyChange(index, 'name', newName)}
                    handlePropertyValueChange={(newValue, index) => this.handleUserDefinedPropertyChange(index, 'value', newValue)}
                    deleteRow={(index) => this.deleteRow(index, true)}
                    reservedPropertyNames={this.state.reservedPropertyNames}
                />
                <form>
                    <div className={leftAlignContainerStyle}>
                        <Button
                            className={buttonStyle}
                            onClick={() => this.addNewProperty(true)}
                        >
                            Add new application specific property
                        </Button>
                    </div>
                </form>
                <hr />
                <form>
                    <FormGroup
                        className={leftAlignContainerStyle}
                        controlId="formControlsMessageBodyText"
                    >
                        <ControlLabel className={headingStyle}>Body</ControlLabel>
                        <FormControl
                            componentClass="textarea"
                            placeholder="Enter message body"
                            className={bodyStyle}
                            value={this.state.messageBody}
                            onChange={(event) => this.handleMessageBodyChange(event.target.value)}
                        />
                    </FormGroup>
                </form>
                <form>
                    <Button
                        id="submitButton"
                        onClick={this.submit}
                    >
                        Submit
                    </Button>
                    <Button
                        onClick={this.discardMessage}
                        bsStyle="danger"
                    >
                        Reset Fields
                    </Button>
                </form>
            </div >
        );
    }

}
