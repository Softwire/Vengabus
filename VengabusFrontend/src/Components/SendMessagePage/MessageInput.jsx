import React, { Component } from 'react';
import { css } from 'react-emotion';
import classNames from 'classnames';
import { MessageBodyInput } from './MessageBodyInput';
import { MessageProperties } from './MessageProperties';
import { MessageDestinationForm } from './MessageDestinationForm';
import { MessageSendAndResetButtons } from './MessageSendAndResetButtons';
import { serviceBusConnection } from '../../AzureWrappers/ServiceBusConnection';
import { cancellablePromiseCollection } from '../../Helpers/CancellablePromiseCollection';
import { sharedSizesAndDimensions, zIndices } from '../../Helpers/SharedSizesAndDimensions';
import _ from 'lodash';
import { UploadMessageFileButton } from '../../Components/UploadMessageFileButton';


const defaultBlankMessage = Object.freeze({
    messageBody: '',
    userDefinedProperties: [],
    preDefinedProperties: []
});
/** 
 * @property {Object} message Can take a message as a prop to replay message.
 * @property {Object} message.customProperties User defined properties (key-string pairs).
 * @property {string} message.messageBody The text of the message.
 * @property {string} message.messageId The ID of the message.
 * @property {string} message.predefinedProperties Any other predefined properties used by Azure.
 * @property {bool} recipientIsQueue Is the recepient of the message a queue?
 * @property {string} selectedQueue Name of the recepient queue.
 * @property {string} selectedTopic Name of the recepient topic.
 */

export class MessageInput extends Component {
    constructor(props) {
        super(props);
        const message = this.props.message;
        this.promiseCollection = new cancellablePromiseCollection();
        this.state = {
            permittedValues: [],
            availableTopics: [],
            availableQueues: [],
            recipientIsQueue: !!(this.props.recipientIsQueue),
            messageBody: message ? message.messageBody : '',
            userDefinedProperties: message ? this.getUserDefinedProperties(message) : [], //[{name: something, value: something}]
            preDefinedProperties: [], //need to fetch permittedValues and reservedPropertyNames before this can be set
            reservedPropertyNames: [], //a list of name of possible readable properties of a message
            selectedQueue: this.props.selectedQueue,
            selectedTopic: this.props.selectedTopic,
            arePreDefinedPropsLoaded: false,
            sendMessageModalWarnings: null
        };
    }

    componentDidMount() {
        this.serviceBusService = serviceBusConnection.getServiceBusService();

        let permittedValuesPromise = this.promiseCollection.addNewPromise(this.serviceBusService.getWriteableMessageProperties());
        let reservedPropertyNamesPromise = this.promiseCollection.addNewPromise(this.serviceBusService.getReadableMessageProperties());
        Promise.all([permittedValuesPromise, reservedPropertyNamesPromise]).then((result) => {
            this.setState({
                permittedValues: result[0],
                reservedPropertyNames: result[1],
                arePreDefinedPropsLoaded: true
            });
            this.setMessageDetails(this.props.messageBody, result[0], result[1]);
        }).catch((e) => { if (!e.isCanceled) { console.log(e); } });


        const fetchQueueDataPromise = this.promiseCollection.addNewPromise(this.serviceBusService.listQueues());
        const fetchTopicDataPromise = this.promiseCollection.addNewPromise(this.serviceBusService.listTopics());
        fetchQueueDataPromise.then((result) => {
            this.setState({
                availableQueues: this.convertArrayOfNamesToValueLabel(result)
            });
        }).catch((e) => { if (!e.isCanceled) { console.log(e); } });
        fetchTopicDataPromise.then((result) => {
            this.setState({
                availableTopics: this.convertArrayOfNamesToValueLabel(result)
            });
        }).catch((e) => { if (!e.isCanceled) { console.log(e); } });
    }

    componentWillUnmount() {
        this.promiseCollection.cancelAllPromises();
    }

    setMessageDetails = (messageObject, permittedValues, reservedValues) => {
        if (!messageObject) {
            return;//this.setState({ ...defaultBlankMessage });
        }
        const preDefinedProps = this.getPreDefinedProperties(messageObject, permittedValues, reservedValues);
        this.setState({
            messageBody: messageObject.messageBody,
            userDefinedProperties: this.getUserDefinedProperties(messageObject),
            preDefinedProperties: preDefinedProps
        });
    }

    setMessageFieldsFromFileObject = (fileReadMessagePromise) => {
        const messagePromise = this.promiseCollection.addNewPromise(fileReadMessagePromise);
        messagePromise.then(messageObjects => {
            this.setMessageDetails(messageObjects[0], this.permittedValues, this.reservedPropertyNames);
        }); //.catch is handled by the LoadFile button - displays a notification. 
    }

    /**
     * Converts a string to an object of the form:
     * `{value: "string", label: "string"}`
     * Used to add values to select elements.
     * @param {string} str The string to convert.
     * @returns {object} The created object.
     */
    convertToValueLabel = (str) => {
        return { value: str, label: str };
    }

    /**
     * Converts an array of objects with a name property to an array of objects with a value and label property:
     * `[{..., name: "example", ...}] -> [{value: "example", label: "example"}]`
     * Used to add values to select elements.
     * @param {object[]} arr The array of objects containing the names to use.
     * @returns {object[]} The created object.
     */
    convertArrayOfNamesToValueLabel = (arr) => {
        return _(arr).map((element) => this.convertToValueLabel(element.name));
    }

    getUserDefinedProperties = (message) => {
        return this.getTargetProperties(message, "customProperties");
    }

    getPreDefinedProperties = (message, permittedValues, reservedPropertyNames) => {
        return this.getTargetProperties(message, "predefinedProperties", permittedValues, reservedPropertyNames);
    }

    /**
     * Gets the properties of a certain type from a message.
     * @param {object} message The message to get the properties from.
     * @param {string} propertyClass The class of properties to get,
     * either `customProperties` or `predefinedProperties`.
     * @param {string[]} settableProps List of keys whose value can be written.
     * @param {string[]} gettableProps List of keys whose value can be read.
     * @returns {object[]} The list of properties.
     */
    getTargetProperties = (message, propertyClass, settableProps, gettableProps) => {
        const outputProperties = [];
        const properties = message[propertyClass];
        const acceptAllProps = !settableProps || settableProps.length === 0;
        if (properties) { //check if properties are defined
            const keys = Object.keys(message[propertyClass]);
            for (let i = 0; i < keys.length; i++) {
                const key = keys[i];
                if (acceptAllProps || settableProps.includes(key)) {
                    outputProperties.push({
                        name: key,
                        value: message[propertyClass][key]
                    });
                } else if (!gettableProps.includes(key)) {  //if the property is read-only then that is expected and can be ignored
                    //if this error is thrown then it is expected to be a programming error, not a user error or a bad message
                    throw new Error(`key ${key} was not an expected predefined property`);
                }
            }
        }
        return outputProperties;
    }

    handlePropertiesChange = (propertyType, newProperties) => {
        this.setState({ [propertyType]: newProperties });
    }

    /**
     * Updates a collection of properties by applying the given updateOperation to it.
     * @param {boolean} isUserDefined Should be true if the property is user-defined, false if it is a pre-defined property.
     * @param {funciton} updateOperation The operation to be applied to the properties collection.
     */
    updatePropertiesCollection = (isUserDefined, updateOperation) => {
        const propertyType = isUserDefined ? "userDefinedProperties" : "preDefinedProperties";
        const newProperties = [...this.state[propertyType]];
        updateOperation(newProperties);
        this.setState({
            //https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names
            [propertyType]: newProperties
        });
    };

    setWarnings = (warnings) => {
        this.propertyWarnings = warnings;
    }

    //generate warnings JSX.
    constructPropertyWarnings = () => {
        if (!this.propertyWarnings || !this.propertyWarnings.length) {
            this.setState({ sendMessageModalWarnings: null });
        } else {
            const warnings = this.propertyWarnings.map((value) => (<p key={"Warning " + value}>{value}</p>));
            this.setState({
                sendMessageModalWarnings: (<React.Fragment>{warnings}</React.Fragment>)
            });
        }
    }

    /**
     * Updates the message body in the state with a new value.
     * @param {string} newBody The new value of the body.
     */
    handleMessageBodyChange = newBody => {
        this.setState({ messageBody: newBody });
    };

    handleDestinationChange = (isDestinationQueue, destinationName) => {
        this.setState({
            recipientIsQueue: isDestinationQueue,
            [isDestinationQueue ? "selectedQueue" : "selectedTopic"]: destinationName
        });
    }

    /**
     * Converts an array of properties to a properties object:
     * [{name: name1, value: value1}] -> {name1: value1}
     * @param {object[]} properties The array of properties to convert.
     * @returns {object} The created properties object.
     */
    createMessagePropertyDictionary = (properties) => {
        const ret = {};
        for (let i = 0; i < properties.length; i++) {
            const thisPropertyName = properties[i].name;
            const thisPropertyValue = properties[i].value;
            //Prevent the user from inputting invalid property names.
            //Cannot use isPropertyNameValid here because if there are two properties with the same name it will mark
            //both of them as invalid whereas we just want to remove one of them.
            if (thisPropertyName && thisPropertyValue && !ret.hasOwnProperty(thisPropertyName)) {
                if (thisPropertyName.length > 0) {
                    ret[thisPropertyName] = thisPropertyValue;
                }
            }
        }
        return ret;
    }

    /**
     * Converts the userDefinedProperties and preDefinedProperties arrays to a single message object
     * in the format that is accepted by the API.
     * @returns {object} The created message.
     */
    createMessageObject = () => {
        const message = {};
        message.customProperties = this.createMessagePropertyDictionary(this.state.userDefinedProperties);
        message.predefinedProperties = this.createMessagePropertyDictionary(this.state.preDefinedProperties);
        message.messageBody = this.state.messageBody;
        return message;
    }

    /**
     * Clears the message body and removes all properties from the message.
     */
    discardMessage = () => {
        this.setState({ ...defaultBlankMessage });
    }

    /**
     * Sends the message to the selected queue/topic.
     *  @returns {nothing} for the promice
     */
    submit = () => {
        const message = this.createMessageObject();
        if (this.state.recipientIsQueue) {
            return this.serviceBusService.sendMessageToQueue(this.state.selectedQueue, message);
        } else {
            return this.serviceBusService.sendMessageToTopic(this.state.selectedTopic, message);
        }
    }

    render() {
        let stickyHeight = "150px";
        const formStyle = css`
            margin-left: 5px;
            margin-right: 5px;
            width: calc(100% - 10px); /* 10px total margin */
            float: left;
        `;
        const fullWidth = css`
            float: left;
            width: 100%;
        `;
        const topSticky = css`
            z-index: ${zIndices.SEND_MESSAGE_STICKY};
            position: fixed;
            width: calc(100% - ${sharedSizesAndDimensions.SIDEBAR_WIDTH}px);
            height: ${stickyHeight};
            background-color: #ECEFF1;
            border-bottom: 2px solid black;
            padding-top: 10px;
            margin-left: -5px; /* Form margin undesirable for sticky section */
            padding-left: 10px;
            padding-right: 10px;
        `;
        const stickySpacer = css`
            height: calc(${stickyHeight} + 10px);
        `;
        const destinationFormStyle = css`
            width: 450px;
            float: left;
        `;
        const vertAlignButtons = css`
            margin-top: 10px;
        `;
        const floatRightWithMargin = css`
            float: right;
            margin-right:20px;
        `;

        const selectedEndpoint = this.state.recipientIsQueue ? this.state.selectedQueue : this.state.selectedTopic;

        //construct child components
        const messageDestinationForm = <MessageDestinationForm
            recipientIsQueue={this.state.recipientIsQueue}
            availableQueues={this.state.availableQueues}
            availableTopics={this.state.availableTopics}
            selectedQueue={this.state.selectedQueue}
            selectedTopic={this.state.selectedTopic}
            handleDestinationChange={this.handleDestinationChange}
        />;

        const mainButtons = <MessageSendAndResetButtons
            selectedEndpoint={selectedEndpoint}
            warnings={this.state.sendMessageModalWarnings}
            generateWarnings={this.constructPropertyWarnings}
            submit={this.submit}
            discardMessage={this.discardMessage}
        />;

        const uploadFromFile = (
            <React.Fragment>
                <p>Upload Message from File</p>

                <UploadMessageFileButton
                    disabled={!this.state.arePreDefinedPropsLoaded}
                    onUpload={this.setMessageFieldsFromFileObject}
                    text="Upload message from file"
                />
            </React.Fragment>
        );

        const messageProperties = <MessageProperties
            arePreDefinedPropsLoaded={this.state.arePreDefinedPropsLoaded}
            preDefinedProperties={this.state.preDefinedProperties}
            userDefinedProperties={this.state.userDefinedProperties}
            permittedValues={this.state.permittedValues}
            reservedPropertyNames={this.state.reservedPropertyNames}
            handlePropertiesChange={this.handlePropertiesChange}
            reportWarnings={this.setWarnings}
        />;

        const messageBodyInput = <MessageBodyInput
            messageBody={this.state.messageBody}
            handleMessageBodyChange={this.handleMessageBodyChange}
        />;

        return (
            <div className={formStyle} >
                <div className={topSticky}>
                    <div className={destinationFormStyle}>
                        {messageDestinationForm}
                    </div>
                    <div className={classNames(vertAlignButtons, floatRightWithMargin)}>
                        {mainButtons}
                    </div>
                </div>
                <div className={stickySpacer} />
                {uploadFromFile}
                <hr className={fullWidth} />
                {messageProperties}
                <hr className={fullWidth} />
                {messageBodyInput}
            </div >
        );
    }

}
