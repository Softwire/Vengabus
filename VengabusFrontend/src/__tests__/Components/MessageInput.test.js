import 'jest-localstorage-mock';
import { configure } from 'enzyme';
import Adaptor from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import React from 'react';
import { MessageInput } from '../../Components/MessageInput';

configure({ adapter: new Adaptor() });

//QQ Inplement a test that checks that the body has been added to the message
//once the submitMessage function actually does something

it('renders correctly', () => {
    let messagePropertyInput = renderer.create(
        <MessageInput />);
    expect(messagePropertyInput.toJSON()).toMatchSnapshot();
});