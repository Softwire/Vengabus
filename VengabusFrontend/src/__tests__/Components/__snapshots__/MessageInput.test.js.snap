// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="css-lushey"
>
  <div
    className="css-mxbhzi"
  >
    <p
      className="css-wni10w"
    >
      Pre-defined Properties
    </p>
  </div>
  <div />
  <form>
    <div
      className="css-mxbhzi"
    >
      <button
        className="css-74801x btn btn-default"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Add new Azure property
      </button>
    </div>
  </form>
  <hr />
  <div
    className="css-mxbhzi"
  >
    <p
      className="css-wni10w"
    >
      User-defined Properties
    </p>
  </div>
  <div />
  <form>
    <div
      className="css-mxbhzi"
    >
      <button
        className="css-74801x btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Add new application specific property
      </button>
    </div>
  </form>
  <hr />
  <form>
    <div
      className="css-mxbhzi form-group"
    >
      <label
        className="css-wni10w control-label"
        htmlFor="formControlsMessageBodyText"
      >
        Body
      </label>
      <textarea
        className="css-cjfa1z form-control"
        id="formControlsMessageBodyText"
        onChange={[Function]}
        placeholder="Enter message body"
        type={undefined}
        value=""
      />
    </div>
  </form>
  <form>
    <button
      className="btn btn-default"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
    <button
      className="btn btn-danger"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Reset Fields
    </button>
  </form>
</div>
`;

exports[`renders correctly from a predefined message 1`] = `
<div
  className="css-lushey"
>
  <div
    className="css-mxbhzi"
  >
    <p
      className="css-wni10w"
    >
      Pre-defined Properties
    </p>
  </div>
  <div />
  <form>
    <div
      className="css-mxbhzi"
    >
      <button
        className="css-74801x btn btn-default"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Add new Azure property
      </button>
    </div>
  </form>
  <hr />
  <div
    className="css-mxbhzi"
  >
    <p
      className="css-wni10w"
    >
      User-defined Properties
    </p>
  </div>
  <div>
    <div>
      <form
        className="css-1rnsx0x"
      >
        <div
          className="form-group"
        >
          <input
            className="css-83yc3x form-control"
            id="formControlsText"
            onChange={[Function]}
            placeholder="Enter property name"
            type="text"
            value="userDefinedProp1"
          />
        </div>
      </form>
      <form
        className="css-mr74sa"
      >
        <div
          className="form-group"
        >
          <input
            className="css-83yc3x form-control"
            id="formControlsText"
            onChange={[Function]}
            placeholder="Enter property value"
            type="text"
            value="value1"
          />
        </div>
      </form>
      <div
        className="css-146pozp"
      >
        <button
          className="css-83yc3x btn btn-danger"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div>
      <form
        className="css-1rnsx0x"
      >
        <div
          className="form-group"
        >
          <input
            className="css-83yc3x form-control"
            id="formControlsText"
            onChange={[Function]}
            placeholder="Enter property name"
            type="text"
            value="userDefinedProp2"
          />
        </div>
      </form>
      <form
        className="css-mr74sa"
      >
        <div
          className="form-group"
        >
          <input
            className="css-83yc3x form-control"
            id="formControlsText"
            onChange={[Function]}
            placeholder="Enter property value"
            type="text"
            value="value2"
          />
        </div>
      </form>
      <div
        className="css-146pozp"
      >
        <button
          className="css-83yc3x btn btn-danger"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
  <form>
    <div
      className="css-mxbhzi"
    >
      <button
        className="css-74801x btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Add new application specific property
      </button>
    </div>
  </form>
  <hr />
  <form>
    <div
      className="css-mxbhzi form-group"
    >
      <label
        className="css-wni10w control-label"
        htmlFor="formControlsMessageBodyText"
      >
        Body
      </label>
      <textarea
        className="css-cjfa1z form-control"
        id="formControlsMessageBodyText"
        onChange={[Function]}
        placeholder="Enter message body"
        type={undefined}
        value="Hello world!"
      />
    </div>
  </form>
  <form>
    <button
      className="btn btn-default"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
    <button
      className="btn btn-danger"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Reset Fields
    </button>
  </form>
</div>
`;
